cloud_tasks_queue: supress permadiffs on backoff settings

wyardley opened the pull request:
Add `tpgresource.DurationDiffSuppress` to min / max backoff
Fixes https://github.com/hashicorp/terraform-provider-google/issues/15166

h/t to @edwardmedia for the pointer

Note: `maxRetryDuration` already has a custom diff function: `suppressOmittedMaxDuration()` - for now, I didn't touch that one

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ ] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
cloud_tasks: suppressed time-unit permadiffs on `google_cloud_tasks_queue` min and max backoff settings
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I've detected that you're a community contributor. @slevenick, a repository maintainer, has been assigned to assist you and help review your changes. 

<details>
  <summary>:question: First time contributing? Click here for more details</summary>

---

Your assigned reviewer will help review your code by: 
* Ensuring it's backwards compatible, covers common error cases, etc.
* Summarizing the change into a user-facing changelog note.
* Passes tests, either our "VCR" suite, a set of presubmit tests, or with manual test runs.

You can help make sure that review is quick by running local tests and ensuring they're passing in between each push you make to your PR's branch. Also, try to leave a comment with each push you make, as pushes generally don't generate emails.

If your reviewer doesn't get back to you within a week after your most recent change, please feel free to leave a comment on the issue asking them to take a look! In the absence of a dedicated review dashboard most maintainers manage their pending reviews through email, and those will sometimes get lost in their inbox.

---

</details>

wyardley commented:
So far, adding `diff_suppress_func` doesn't seem to modify the resource code when I generate providers; not sure if that's expected. Working on getting setup to run lint / unit tests locally again, but added a simple copy / paste test that I think will test for the issue?

slevenick commented:
> So far, adding `diff_suppress_func` doesn't seem to modify the resource code when I generate providers; not sure if that's expected. Working on getting setup to run lint / unit tests locally again, but added a simple copy / paste test that I think will test for the issue?

It *should* cause a change in the provider, I'd guess you don't have the local environment set up correctly. I've triggered the run so it should show the diff here in ~15 minutes

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 48 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 48 insertions(+), 6 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `13` 
Passed tests `13` 
Skipped tests: `0` 
Affected tests: `0` 

$\textcolor{red}{\textsf{Errors occurred during REPLAYING mode. Please fix them to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b6439791-db5c-4dd4-8d44-33e1e025a2aa/build-log/replaying_test.log)

wyardley commented:
> Your test config function doesn't match up to what is being called.

Derp, yeah, sorry, updated in one place but not the other. updated.

Thoughts on whether we need to solve this in `maxRetryDuration` too? I assume it's not trivial to add two functions, but also seems janky to copy / paste into the existing custom function?

slevenick commented:
> > Your test config function doesn't match up to what is being called.
> 
> Derp, yeah, sorry, updated in one place but not the other. updated.
> 
> Thoughts on whether we need to solve this in `maxRetryDuration` too? I assume it's not trivial to add two functions, but also seems janky to copy / paste into the existing custom function?

Judging by the name that looks to solve a different problem. Do you see this same problem on maxRetryDuration?

The existing code in that DSF just equalizes `"0"` and `""`

wyardley commented:
> I'd guess you don't have the local environment set up correctly.

Yeah, possible.

I was generating _some_ diffs, just not the expected ones.

I have successfully done this before, and have run through all the setup docs in the past, and reviewed them again. The doctor script didn't report any problems. Still dealing with some go dependency issues for running vet / lint / etc., but this should be happening before that, I think. Anyway, I'll get it working again, esp. when I next dig into anything trickier than this.

wyardley commented:
> Judging by the name that looks to solve a different problem. Do you see this same problem on maxRetryDuration?

That's right - the existing custom code (and test) seems to be handling the situation where it's 0. Since it accepts time units, I guess I was expecting that the same issue would likely affect `maxRetryDuration` -- it hasn't come up for me specifically, but it would seem like it _could_?

slevenick commented:
> > Judging by the name that looks to solve a different problem. Do you see this same problem on maxRetryDuration?
> 
> That's right - the existing custom code (and test) seems to be handling the situation where it's 0. Since it accepts time units, I guess I was expecting that the same issue would likely affect `maxRetryDuration` -- it hasn't come up for me specifically, but it would seem like it _could_?
> 
> I am not suggesting that we remove the existing custom function for it, just that we might _also_ ideally want to incorporate the behavior of `tpgresource.DurationDiffSuppress` as well, if you follow.

Yeah, we could incorporate the behavior in there by calling the duration DSF. I'd make sure we actually see the problem before implementing that fix though

wyardley commented:
@slevenick 

Yes, as you'd guess, the same behavior affects `max_retry_duration` - see the example below

```hcl
resource "google_cloud_tasks_queue" "foobarbaz" {
  name     = "foobarbaz"
  location = "us-west2"
  rate_limits {
    max_dispatches_per_second = 8
    max_concurrent_dispatches = 2
  }   
  retry_config {
    max_attempts = 2
    max_backoff = "2.5s"
    min_backoff = "2.000s"
    max_doublings = 5
    max_retry_duration = "10.0s"
  }
}    
```

will result in this permadiff
```
      ~ retry_config {
          ~ max_backoff        = "2.500s" -> "2.5s"
          ~ max_retry_duration = "10s" -> "10.0s"
          ~ min_backoff        = "2s" -> "2.000s"
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 48 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 48 insertions(+), 6 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2855` 
Passed tests `2555` 
Skipped tests: `297` 
Affected tests: `3` 

#### Action taken 
<details> <summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerAwsNodePool_BetaBasicHandWritten|TestAccComputeFirewallPolicyRule_multipleRules|TestAccCloudTasksQueue_TimeUnitDiff </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccCloudTasksQueue_TimeUnitDiff`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/recording/TestAccCloudTasksQueue_TimeUnitDiff.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccContainerAwsNodePool_BetaBasicHandWritten`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/build-log/recording_build/TestAccContainerAwsNodePool_BetaBasicHandWritten_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/recording/TestAccContainerAwsNodePool_BetaBasicHandWritten.log)]
`TestAccComputeFirewallPolicyRule_multipleRules`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/build-log/recording_build/TestAccComputeFirewallPolicyRule_multipleRules_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/recording/TestAccComputeFirewallPolicyRule_multipleRules.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/2593006a-c775-43ae-9c20-b1b505079936/recording) for each test

wyardley commented:
The test as written will already check that there's no drift after apply, right? Or do I have to do the thing where it does it a second time?

> Yeah, we could incorporate the behavior in there by calling the duration DSF

I can add `max_retry_duration` to the existing test (which I think should then fail); do you have an example or suggestion of how I'd do that? Just call `tpgresource.DurationDiffSuppress()` within `suppressSuffixDiff()`?

slevenick commented:
> The test as written will already check that there's no drift after apply, right? Or do I have to do the thing where it does it a second time?
> 
> > Yeah, we could incorporate the behavior in there by calling the duration DSF
> 
> I can add `max_retry_duration` to the existing test (which I think should then fail); do you have an example or suggestion of how I'd do that? Just call `tpgresource.DurationDiffSuppress()` within `suppressSuffixDiff()`?

The test should check there's no diff after apply. You can test that out by changing the max_retry_duration to a value that causes the issue.

I think you should be able to call the DurationDiffSuppress from within the existing handwritten diff suppress that field is using, just by replacing:
```
return false
```
with
```
return tpgresource.DurationDiffSuppress(ARGS HERE)
```

wyardley commented:
Sounds good. I can't run the integration tests locally, so I'll push up the change with the presumably breaking test.

Assuming that looks good (bad), I'll push up the change with your suggestion for calling the additional suppression function; thanks for the example.

wyardley commented:
So like this?

```go
diff --git a/mmv1/templates/terraform/constants/cloud_tasks_retry_config_custom_diff.go b/mmv1/templates/terraform/constants/cloud_tasks_retry_config_custom_diff.go
index cba219445..9f65d60e8 100644
--- a/mmv1/templates/terraform/constants/cloud_tasks_retry_config_custom_diff.go
+++ b/mmv1/templates/terraform/constants/cloud_tasks_retry_config_custom_diff.go
@@ -1,7 +1,11 @@
+import (
+        "github.com/GoogleCloudPlatform/terraform-google-conversion/v2/tfplan2cai/converters/google/resources/tpgresource"
+)
+
 func suppressOmittedMaxDuration(_, old, new string, _ *schema.ResourceData) bool {
 	if old == "" && new == "0s" {
 		log.Printf("[INFO] max retry is 0s and api omitted field, suppressing diff")
 		return true
 	}
-	return false
+	return tpgresource.DurationDiffSuppress(old, new)
 }
```

wyardley commented:
@slevenick ok, I had to pass the other two parameters too, but seems to work (and confirmed that the import is already there / not needed)! 🎉 
I just used `k` and `d` for the var names, let me know if I need to use something more descriptive.

I fixed the local build and confirms that this works locally to resolve the issue.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 50 insertions(+), 8 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 50 insertions(+), 8 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8359-old..auto-pr-8359) ( 3 files changed, 5 insertions(+), 5 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2864` 
Passed tests `2562` 
Skipped tests: `299` 
Affected tests: `3` 

#### Action taken 
<details> <summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccCloudTasksQueue_TimeUnitDiff|TestAccComputeFirewallPolicyRule_multipleRules|TestAccContainerAwsNodePool_BetaBasicHandWritten </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccCloudTasksQueue_TimeUnitDiff`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/recording/TestAccCloudTasksQueue_TimeUnitDiff.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccComputeFirewallPolicyRule_multipleRules`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/build-log/recording_build/TestAccComputeFirewallPolicyRule_multipleRules_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/recording/TestAccComputeFirewallPolicyRule_multipleRules.log)]
`TestAccContainerAwsNodePool_BetaBasicHandWritten`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/build-log/recording_build/TestAccContainerAwsNodePool_BetaBasicHandWritten_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/recording/TestAccContainerAwsNodePool_BetaBasicHandWritten.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/f0647150-72f6-425c-a4ed-2e6f948006db/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 50 insertions(+), 8 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8359-old..auto-pr-8359) ( 2 files changed, 50 insertions(+), 8 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8359-old..auto-pr-8359) ( 3 files changed, 5 insertions(+), 5 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2872` 
Passed tests `2571` 
Skipped tests: `299` 
Affected tests: `2` 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccContainerAwsNodePool_BetaBasicHandWritten|TestAccComputeFirewallPolicyRule_multipleRules </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccContainerAwsNodePool_BetaBasicHandWritten`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/build-log/recording_build/TestAccContainerAwsNodePool_BetaBasicHandWritten_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/recording/TestAccContainerAwsNodePool_BetaBasicHandWritten.log)]
`TestAccComputeFirewallPolicyRule_multipleRules`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/build-log/recording_build/TestAccComputeFirewallPolicyRule_multipleRules_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/recording/TestAccComputeFirewallPolicyRule_multipleRules.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8359/artifacts/b65d9c2a-7b04-4737-8a2e-415f59aad837/recording) for each test

wyardley commented:
@slevenick let me know how this looks now
I think the affected tests are not related, right?
